### ISSUE TITLE
[Feature request] Incompatability with `subpar` package

### ISSUE THREAD

**Issue by virgiling** [RubixDev/typst-i-figured#12]
Hi there! I have been using the package [subpar](https://github.com/tingerrr/subpar) to draw sub-figures, but when I combined `i-figured` with `subpar`,  references went wrong.

As an example, I wrote code like this:

```typ
#import "@preview/subpar:0.1.1": grid as subfigure

We can use package `subpar` to draw sub-figures, which can be referenced as @fig:ida-star, and sub-figures @fig:ida-star-1, @fig:ida-star-2

#subfigure(
  figure(
    image("fig/ida-star-1.png", width: 50%),
    caption: [first step],
  ),
  <ida-star-1>,
  figure(
    image("fig/ida-star-2.png", width: 50%),
    caption: [second step],
  ),
  <ida-star-2>,
  columns: (1fr, 1fr),
  caption: [usage of package `subpar`],
  numbering-sub: "a",
  label: <ida-star>,
)
```
and the references wound be:
![image](https://github.com/user-attachments/assets/b0dd4524-1fe1-4557-aebc-24fdc14d3cb3)

It seems that this idea is not compatible with your package and I am not sure it is due to the two packages not playing well together.

**Comment by Rsweater** [comment]
Friend, hello~ Have you found a solution to this problem? It's frustrating when there are numbering issues even without using the `subpar` package:
`#figure(grid(columns: 2, gutter: 1em,
  figure(rect(), numbering: none, caption: [a) demo1]),
  figure(rect(), numbering: none, caption: [b) demo2]),
), caption: [demo]) <fig:demo>

@fig:demo (a) is xxx`
However, `numbering` does not work, and the effect is as follows:
![局部截取_20241223_141854](https://github.com/user-attachments/assets/444d2008-dbc7-49d5-b8d9-ac89c11bf904)

**Comment by virgiling** [comment]
> Friend, hello~ Have you found a solution to this problem? It's frustrating when there are numbering issues even without using the `subpar` package: `#figure(grid(columns: 2, gutter: 1em, figure(rect(), numbering: none, caption: [a) demo1]), figure(rect(), numbering: none, caption: [b) demo2]), ), caption: [demo]) fig:demo
> 
> @fig:demo (a) is xxx`However,`numbering` does not work, and the effect is as follows: ![局部截取_20241223_141854](https://private-user-images.githubusercontent.com/49091147/398070069-444d2008-dbc7-49d5-b8d9-ac89c11bf904.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ5NTM5NzAsIm5iZiI6MTczNDk1MzY3MCwicGF0aCI6Ii80OTA5MTE0Ny8zOTgwNzAwNjktNDQ0ZDIwMDgtZGJjNy00OWQ1LWI4ZDktYWM4OWMxMWJmOTA0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjIzVDExMzQzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNjYmM3OWNmN2NkMDJjMDQxMGJjNjJhODg2YTBlMWJhMjk3NjNlMGMwZGY3MDhjZTQyYTlmNDcxMjRhMTlmNGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.p3l4YbKTYan3Xyzp4LHyWzLnP5EU5eh0coTQsW-22o0)

Unfortunately, I haven't found a solution yet, but I feel that `subpar` could be rewritten to fit this package. I will give it a try when I'm free.

**Comment by Rsweater** [comment]
it's good. I can have a try when i have free time.

**Comment by Rsweater** [comment]
@topdeoo  It seems that the `i-figured` count also has an issue [code](https://github.com/RubixDev/typst-i-figured/blob/36e238538142ed123d6493a49d756088587d075b/i-figured.typ#L33), which will directly ignore numbering. 
But due to limited capabilities, I don't know how to modify it.

It seems that the reason why `i-figured` is incompatible with `subpar` is that the subplots in `subpar` have a new `kind`, while `i-figured` uses a new `kind` to distinguish eqt, fig, etc.
```typst
#figure(grid(columns: 2, gutter: 1em,
  figure(rect(), numbering: none, caption: [a) demo1]),
  figure(rect(), numbering: none, caption: [b) demo2]),
), caption: [demo]) <fig:demo>

@fig:demo (a) is xxx
``` 
This piece of code does not use `numbering` (the disadvantage is that subfigures cannot be referenced, but this is not the focus of discussion now), and the display result is as follows: 
![ef8e665467d84bbe33b8295b3ac93abc](https://github.com/user-attachments/assets/0473bad0-9bce-4d47-873d-02af4cf7e7c3)

```typst
#import "@preview/i-figured:0.2.4"
#show figure: i-figured.show-figure

#figure(grid(columns: 2, gutter: 1em,
  figure(rect(), numbering: none, caption: [a) demo1]),
  figure(rect(), numbering: none, caption: [b) demo2]),
), caption: [demo]) <fig:demo>

@fig:demo (a) is xxx
``` 
The effect of the above code is as follows:
![13a89a60d54bf5eedc0ba34ebeeb7f6c](https://github.com/user-attachments/assets/0bb14882-736e-44e1-856a-1f03186f20c7)

Using the `subpar` package alone yields normal results, as shown in the code and effect image:
```
#import "@preview/subpar:0.2.0": grid as subfigure
#subfigure(
  figure(rect(), caption: [
    An image of the andromeda galaxy.
  ]), <a>,
  figure(rect(), caption: [
    A sunset illuminating the sky above a mountain range.
  ]), <b>,
  columns: (1fr, 1fr),
  caption: [A figure composed of two sub figures.],
  label: <full>,
)

*subpar displays*: Above in @full, we see a figure which is composed of two other figures, namely @a and @b.
```
![局部截取_20241225_173722](https://github.com/user-attachments/assets/da6a7e95-2ab0-4d7b-832c-02460d1db689)

**Comment by virgiling** [comment]
> ```typst
> #import "@preview/i-figured:0.2.4"
> #show figure: i-figured.show-figure
> 
> #figure(grid(columns: 2, gutter: 1em,
>   figure(rect(), numbering: none, caption: [a) demo1]),
>   figure(rect(), numbering: none, caption: [b) demo2]),
> ), caption: [demo]) <fig:demo>
> 
> @fig:demo (a) is xxx
> ```

Yeah, the solution seems to be add a new kind `subfigure` in this package, but `subfigure` should share the counter with other figure whose kind is `i-figured-figure`. In the mean time, when call `subfigure`, the kind (or counter?) of figure inside should be local(not `figure` since this package will change it into `i-figured-figure`).

**Comment by Rsweater** [comment]
> > ```typst
> > #import "@preview/i-figured:0.2.4"
> > #show figure: i-figured.show-figure
> > 
> > #figure(grid(columns: 2, gutter: 1em,
> >   figure(rect(), numbering: none, caption: [a) demo1]),
> >   figure(rect(), numbering: none, caption: [b) demo2]),
> > ), caption: [demo]) <fig:demo>
> > 
> > @fig:demo (a) is xxx
> > ```
> 
> Yeah, the solution seems to be add a new kind `subfigure` in this package, but `subfigure` should share the counter with other figure whose kind is `i-figured-figure`. In the mean time, when call `subfigure`, the kind (or counter?) of figure inside should be local(not `figure` since this package will change it into `i-figured-figure`).

Yeah, that's the case. There are quite a few changes to be made, basically merging two packages into one. Either add `fig` and `eqt` subclasses to `subpar` to re-count and add chapter prefixes (the purpose of `i-figured`), or add sub-figure types and sub-figure counters to `i-figured` (the purpose of `subpar`). 
A bit at a loss, this will lead to the direct replacement of these two packages~

**Comment by Rsweater** [comment]
@topdeoo Hi, Conflicts can be resolved in this way：
```typst
// Synopsis:
// - adding contextual numbering like chapter-relative numbering preserves the correct subfigure
//   numbering and supplements

#let fake-image = rect()
#import "@preview/subpar:0.2.0"

#let sub-figure-numbering = (super, sub) => numbering("1.1a", counter(heading).get().first(), super, sub)
#let figure-numbering = super => numbering("1.1", counter(heading).get().first(), super)
#let equation-numbering = super => numbering("(1.1)", counter(heading).get().first(), super)

#set heading(numbering: "1.1")
#show heading.where(level: 1): it => {
  counter(math.equation).update(0)
  counter(figure.where(kind: image)).update(0)
  counter(figure.where(kind: table)).update(0)
  counter(figure.where(kind: raw)).update(0)
  it
}
#show figure.caption: it => {
  let pattern = "^[^:]+" + sym.space.nobreak + "[\d.]+"
  show regex(pattern): strong
  show regex(pattern): emph
  // show regex(pattern): set text(weight: "bold")
  // show regex(pattern): set text(style: "italic")
  it
}

#show figure: set figure(numbering: figure-numbering)
#show math.equation: set math.equation(numbering: equation-numbering)

#let subpar-grid = subpar.grid.with(
  numbering: figure-numbering,
  numbering-sub-ref: sub-figure-numbering,
)

#outline(target: figure.where(kind: image))
#outline(target: figure.where(kind: table))

= Chapter

#figure(
  table(
    columns: 4,
    stroke: none,
    table.hline(),
    [t], [1], [2], [3],
    table.hline(stroke: .5pt),
    [y], [0.3s], [0.4s], [0.8s],
    table.hline(),
  ),
  caption: [三线表],
) <timing-tlt>

$ phi.alt := (1 + sqrt(5)) / 2 $ <ratio>

#figure(fake-image, caption: [aaa])

#subpar-grid(
  figure(fake-image, caption: [Inner caption]), <a>,
  figure(fake-image, caption: [Inner caption]), <b>,
  columns: (1fr, 1fr),
  caption: [Outer caption],
  label: <full1>,
)

#figure(fake-image, caption: [aaa])

#subpar-grid(
  figure(`adas`, caption: [Inner caption]), <c>,
  figure(fake-image, caption: [Inner caption]), <d>,
  columns: (1fr, 1fr),
  caption: [Outer caption],
  label: <full2>,
)

= Another Chapter
$ phi.alt := (1 + sqrt(5)) / 2 $ <ratio2>
#figure(
  table(
    columns: 4,
    stroke: none,
    table.hline(),
    [t], [1], [2], [3],
    table.hline(stroke: .5pt),
    [y], [0.3s], [0.4s], [0.8s],
    table.hline(),
  ),
  caption: [三线表],
) <timing-tlt1>

#figure(fake-image, caption: [aaa])

See @full1, @a and @b.

See also @full2, @c and @d.
```
![局部截取_20250116_040540](https://github.com/user-attachments/assets/86b8e5f8-25e2-4555-96ba-fc37c1e21a03)

**Comment by virgiling** [comment]
> @topdeoo Hi, Conflicts can be resolved in this way：
> 
> ```typst
> // Synopsis:
> // - adding contextual numbering like chapter-relative numbering preserves the correct subfigure
> //   numbering and supplements
> 
> #let fake-image = rect()
> #import "@preview/subpar:0.2.0"
> 
> #let sub-figure-numbering = (super, sub) => numbering("1.1a", counter(heading).get().first(), super, sub)
> #let figure-numbering = super => numbering("1.1", counter(heading).get().first(), super)
> #let equation-numbering = super => numbering("(1.1)", counter(heading).get().first(), super)
> 
> #set heading(numbering: "1.1")
> #show heading.where(level: 1): it => {
>   counter(math.equation).update(0)
>   counter(figure.where(kind: image)).update(0)
>   counter(figure.where(kind: table)).update(0)
>   counter(figure.where(kind: raw)).update(0)
>   it
> }
> #show figure.caption: it => {
>   let pattern = "^[^:]+" + sym.space.nobreak + "[\d.]+"
>   show regex(pattern): strong
>   show regex(pattern): emph
>   // show regex(pattern): set text(weight: "bold")
>   // show regex(pattern): set text(style: "italic")
>   it
> }
> 
> #show figure: set figure(numbering: figure-numbering)
> #show math.equation: set math.equation(numbering: equation-numbering)
> 
> #let subpar-grid = subpar.grid.with(
>   numbering: figure-numbering,
>   numbering-sub-ref: sub-figure-numbering,
> )
> 
> #outline(target: figure.where(kind: image))
> #outline(target: figure.where(kind: table))
> 
> = Chapter
> 
> #figure(
>   table(
>     columns: 4,
>     stroke: none,
>     table.hline(),
>     [t], [1], [2], [3],
>     table.hline(stroke: .5pt),
>     [y], [0.3s], [0.4s], [0.8s],
>     table.hline(),
>   ),
>   caption: [三线表],
> ) <timing-tlt>
> 
> $ phi.alt := (1 + sqrt(5)) / 2 $ <ratio>
> 
> #figure(fake-image, caption: [aaa])
> 
> #subpar-grid(
>   figure(fake-image, caption: [Inner caption]), <a>,
>   figure(fake-image, caption: [Inner caption]), <b>,
>   columns: (1fr, 1fr),
>   caption: [Outer caption],
>   label: <full1>,
> )
> 
> #figure(fake-image, caption: [aaa])
> 
> #subpar-grid(
>   figure(`adas`, caption: [Inner caption]), <c>,
>   figure(fake-image, caption: [Inner caption]), <d>,
>   columns: (1fr, 1fr),
>   caption: [Outer caption],
>   label: <full2>,
> )
> 
> = Another Chapter
> $ phi.alt := (1 + sqrt(5)) / 2 $ <ratio2>
> #figure(
>   table(
>     columns: 4,
>     stroke: none,
>     table.hline(),
>     [t], [1], [2], [3],
>     table.hline(stroke: .5pt),
>     [y], [0.3s], [0.4s], [0.8s],
>     table.hline(),
>   ),
>   caption: [三线表],
> ) <timing-tlt1>
> 
> #figure(fake-image, caption: [aaa])
> 
> See @full1, @a and @b.
> 
> See also @full2, @c and @d.
> ```
> 
> ![局部截取_20250116_040540](https://private-user-images.githubusercontent.com/49091147/403568455-86b8e5f8-25e2-4555-96ba-fc37c1e21a03.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzcwMjU0MTYsIm5iZiI6MTczNzAyNTExNiwicGF0aCI6Ii80OTA5MTE0Ny80MDM1Njg0NTUtODZiOGU1ZjgtMjVlMi00NTU1LTk2YmEtZmMzN2MxZTIxYTAzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTE2VDEwNTgzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI1ZWRiN2Q2NWFjMWFhOGQ1YTE2YTdmNWM4NDUxYWNjNjQzMGY4NDBkNWU1ZjY2YjE1MzEyZTBiMTlkZDE0NzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6_W-CjqwcXAGNz9akgvVg7zxGxgfyktKo3OdK1GWj4s)

Thanks for your solution! 

Since this issue have been solved, I would close it.
